Add support for literate programming.

al45tair opened the pull request:
Adds support for Markdown, reStructuredText and LaTeX files as Swift sources.

al45tair committed:
Add support for literate programming.

Adds support for Markdown, reStructuredText and LaTeX files as Swift sources.

al45tair commented:
See apple/swift#39305

al45tair commented on Sources/PackageLoading/TargetSourcesBuilder.swift:
```diff
@@ -397,6 +400,9 @@ public struct TargetSourcesBuilder {
             }
 
+            // Ignore README files.
+            if path.basename.hasPrefix("README") { continue}
```
I'm not a huge fan of this; I added it because not having it here broke the build (because of a package that was written in C++ but had a `README.md` file in the source directory).

neonichu replied:
Don't think that it is clear cut that readme files in a target are never supposed to be resources and as such ignoring them doesn't seem right. I am curious how these files could have broken the build, unhandled files should only emit warnings?

neonichu replied:
Oh, I see, it is because we're saying `.md` files are now sources, but if they aren't specially formatted, they will break the build?

al45tair replied:
No. It's because there's a package written in C++ that has a `README.md` in its directory, so the package manager code refuses to build it because it thinks it is a mixed-language package.

The literate stuff doesn't care at all about whether or not there is _actually_ code in a Markdown file.

I'm not certain what the right fix is here, though I _am_ certain that this is particular "fix" is a hack and we should do something else.

neonichu replied:
Oh, I see, because `.md` is becoming a source file type, it is seen as mixed source. 

Would it make sense to have a more specific file extension for the literate programming stuff, so that we can distinguish them from "regular" markdown files? In principle, it could be correct that it is mixed source if the .md file *does* contain code.

Another approach could be only treating .md files as source if they're part of a target that isn't a C-family one, but that seems awfully specific.

abertelrud replied:
That would make sense to me.  Using plain `.md` as files to be compiled seems that it would surprise a lot of users.

mattt commented on Sources/PackageLoading/TargetSourcesBuilder.swift:
```diff
@@ -397,6 +400,9 @@ public struct TargetSourcesBuilder {
             }
 
+            // Ignore README files.
+            if path.basename.hasPrefix("README") { continue}
```
Should we accommodate differences in case (e.g. `readme.md`)?

```suggestion
            if path.basenameWithoutExt.uppercased() == "README" { continue }
```

al45tair replied:
Maybe. I'm a bit uncertain that this is the right thing to do in any way whatsoever. Maybe it would be better to detect cases where there are C(++)/ObjC(++) sources and ignore any of the file formats we support for literate Swift.

mattt commented on Sources/PackageLoading/TargetSourcesBuilder.swift:
```diff
@@ -397,6 +400,9 @@ public struct TargetSourcesBuilder {
             }
 
+            // Ignore README files.
+            if path.basename.hasPrefix("README") { continue}
```
Alternatively, to also accommodate variations like `readme-en.md`

```suggestion
            if path.basename.uppercased().hasPrefix("README") { continue }
```

neonichu commented on Sources/PackageLoading/TargetSourcesBuilder.swift:
```diff
@@ -529,7 +535,7 @@ public struct FileRuleDescription {
             rule: .compile,
             toolsVersion: .minimumRequired,
-            fileTypes: ["swift"]
+            fileTypes: ["swift", "md", "rst", "tex"]
```
The new file types should get their own distinct rule with tools-version being 5.6

al45tair committed:
Add literate Swift tests.

Check that literate Swift sources are passed through properly.

al45tair committed:
Turn literate detection off by default, add a target option for it.

Add a `literate` argument to the various target types in the package
description, set by default to false.

Also, widen the existing test for `README` to include a range of other
common all-caps names for files we probably should never compile code
from.

Plus some tests for the above.

assisted-by: Claude
